File: prim.cpp
#include <iostream>

using namespace std;

const int MAX_VERTICES = 100;
const int INF = 1e9;

class Prim {
private:
    int graph[MAX_VERTICES][MAX_VERTICES];
    int numVertices;

public:
    Prim(int n) : numVertices(n) {
        // Initialize the graph with INF (representing absence of edge)
        for (int i = 0; i < numVertices; ++i) {
            for (int j = 0; j < numVertices; ++j) {
                graph[i][j] = INF;
            }
        }
    }

    void addEdge(int u, int v, int weight) {
        graph[u][v] = weight;
        graph[v][u] = weight; // Assuming the graph is undirected
    }

    void primMST() {
        int parent[MAX_VERTICES];
        int key[MAX_VERTICES];
        bool inMST[MAX_VERTICES];

        // Initialize keys and MST set
        for (int i = 0; i < numVertices; ++i) {
            key[i] = INF;
            inMST[i] = false;
        }

        // Start with the first vertex
        key[0] = 0;
        parent[0] = -1;

        for (int count = 0; count < numVertices - 1; ++count) {
            int u = minKey(key, inMST);
            inMST[u] = true;

            for (int v = 0; v < numVertices; ++v) {
                if (graph[u][v] != INF && !inMST[v] && graph[u][v] < key[v]) {
                    parent[v] = u;
                    key[v] = graph[u][v];
                }
            }
        }

        printMST(parent);
    }

    int minKey(const int key[], const bool inMST[]) {
        int minIndex, minValue;
        minIndex = minValue = INF;

        for (int v = 0; v < numVertices; ++v) {
            if (!inMST[v] && key[v] < minValue) {
                minValue = key[v];
                minIndex = v;
            }
        }

        return minIndex;
    }

    void printMST(const int parent[]) {
        cout << "Edges in the Minimum Spanning Tree:" << endl;
        for (int i = 1; i < numVertices; ++i) {
            cout << parent[i] << " - " << i << " : " << graph[i][parent[i]] << endl;
        }
    }
};

int main() {
    int numVertices = 5;

    Prim prim(numVertices);

    prim.addEdge(0, 1, 2);
    prim.addEdge(0, 3, 6);
    prim.addEdge(1, 2, 3);
    prim.addEdge(1, 3, 8);
    prim.addEdge(1, 4, 5);
    prim.addEdge(2, 4, 7);
    prim.addEdge(3, 4, 9);

    prim.primMST();

    return 0;
}

File: prim.dart
class Graph {
  late int vertices;
  late List<List<int>> matrix;

  Graph(int vertices) {
    this.vertices = vertices;
    matrix = List.generate(vertices, (i) => List<int>.filled(vertices, 0));
  }

  void addEdge(int source, int destination, int weight) {
    matrix[source][destination] = weight;
    matrix[destination][source] = weight;
  }

  void primMST() {
    List<bool> inMST = List.filled(vertices, false);
    List<int> key = List.filled(vertices, 999999); // Use a large value instead of infinity
    List<int> parent = List.filled(vertices, -1);

    // Start with the first vertex
    key[0] = 0;

    for (int count = 0; count < vertices - 1; count++) {
      int u = minKey(key, inMST);
      inMST[u] = true;

      for (int v = 0; v < vertices; v++) {
        if (matrix[u][v] != 0 && !inMST[v] && matrix[u][v] < key[v]) {
          parent[v] = u;
          key[v] = matrix[u][v];
        }
      }
    }

    printMST(parent);
  }

  int minKey(List<int> key, List<bool> inMST) {
    int min = 999999; // Use a large value instead of infinity
    int minIndex = -1;

    for (int v = 0; v < vertices; v++) {
      if (!inMST[v] && key[v] < min) {
        min = key[v];
        minIndex = v;
      }
    }

    return minIndex;
  }

  void printMST(List<int> parent) {
    print("Edge \tWeight");
    for (int i = 1; i < vertices; i++) {
      print("${parent[i]} - $i \t${matrix[i][parent[i]]}");
    }
  }
}

void main() {
  Graph graph = Graph(5);

  graph.addEdge(0, 1, 2);
  graph.addEdge(0, 3, 6);
  graph.addEdge(1, 2, 3);
  graph.addEdge(1, 3, 8);
  graph.addEdge(1, 4, 5);
  graph.addEdge(2, 4, 7);
  graph.addEdge(3, 4, 9);

  graph.primMST();
}
